### ISSUE TITLE
Messes up JSON on "Sort and Prettify"

### ISSUE THREAD

**Issue by janosbinder** [federomero/pretty-json#49]
If I do "Sort and Prettify" on this JSON

```javascript
{
  "DV": [
    {
      "RQ": null,
      "ZW": [],
      "BC": 100.0,
      "PopulationFrequencyGeneral": 29.716117216117215,
      "PopulationFrequencyEthnicBackground": 20.316622691292874,
      "Chromosome": "chr22"
    }]
}
```

I get this:

```javascript
{
  "DV": [
    {
      "BC": 100,
      "Chromosome": "chr22",
      "PopulationFrequencyEthnicBackground": {
        "c": [
          2,
          0,
          3,
          1,
          6,
          6,
          2,
          2,
          6,
          9,
          1,
          2,
          9,
          2,
          8,
          7,
          4
        ],
        "e": 1,
        "s": 1
      },
      "PopulationFrequencyGeneral": {
        "c": [
          2,
          9,
          7,
          1,
          6,
          1,
          1,
          7,
          2,
          1,
          6,
          1,
          1,
          7,
          2,
          1,
          5
        ],
        "e": 1,
        "s": 1
      },
      "RQ": null,
      "ZW": [
      ]
    }
  ]
}
```

This must be clearly a bug.

**Comment by janosbinder** [comment]
Atom Version: 1.8.0
pretty-json: 1.4.1
On Ubuntu 16.04

**Comment by lexicalunit** [comment]
Yikes! Looking into this now.

**Comment by lexicalunit** [comment]
So it looks like this is due to the use of [`json-bigint`](https://www.npmjs.com/package/json-bigint) which in turn depends on [`bignumber`](https://www.npmjs.com/package/bignumber.js) which is parsing `29.716117216117215` as a `BigNumber` datatype. When subsequently converted back to a textual representation this is being stringified as an Object type with keys `c`, `e`, and `s` corresponding respectively to _coefficient_, _exponent_, and _sign_.

The reason `Sort and Prettify` causes this to happen but `Prettify` alone doesn't is because the `Sort and Prettify` command uses the `json-stable-stringify` library to satisfy the sorting constraint. This library is unaware of the proper way to stringify a `BigNumber` datatype, and thus you get the default object stringification. The `Prettify` command uses `json-bigint.stringify()` directly and thus everything is properly stringified.

Well. At least now we understand the issue. It's not immediately clear how to resolve it. For now, @janosbinder, the `Prettify` command should at least format your JSON without messing up the numbers. I'm going to keep looking for a way to resolve this as soon as possible.

**Comment by lexicalunit** [comment]
I've implemented a solution in PR #50 that uses the `replacer` option in `json-stable-stringify` to effectively teach it how to handle `BigNumber` datatypes correctly.  I also added a test using the data @janosbinder put in this ticket to ensure no future regressions. Once the PR passes CI tests I'll merge it and publish a new patch version.

Thanks for the report @janosbinder!

**Comment by lexicalunit** [comment]
Published in v1.4.2.
